packages/ses/src/error/note-log-args.js

```diff
@@ -0,0 +1,216 @@
+ */
+
+/**
+ * Makes a new self-linked cell. There are two reasons to do so:
+ *    * To make the head sigil of a new initially-empty list
+ *    * To make a non-sigil cell to be `spliceAfter`ed.
+ *
+ * @param {object} key
+ * @param {any} value
+ * @returns {DoublyLinkedCell}
+ */
+const makeSelfCell = (key, value) => {
+  /** @type {DoublyLinkedCell} */
+  const selfCell = {
+    // @ts-expect-error will be fixed before return
+    next: undefined,
+    // @ts-expect-error will be fixed before return
+    prev: undefined,
+    key,
+    value,
+  };
+  selfCell.next = selfCell;
+  selfCell.prev = selfCell;
+  // Not frozen!
+  return selfCell;
+};
+
+/**
+ * Splices a self-linked non-sigil cell into a list after `prev`.
+ * `prev` could be the head sigil, or it could be some other non-sigil
+ * cell within a list.
+ *
+ * @param {DoublyLinkedCell} prev
+ * @param {DoublyLinkedCell} selfCell
+ */
+const spliceAfter = (prev, selfCell) => {
+  if (prev === selfCell) {
+    throw TypeError('Cannot splice a cell into itself');
+  }
+  if (selfCell.next !== selfCell || selfCell.prev !== selfCell) {
+    throw TypeError('Expected self-linked cell');
+  }
+  const cell = selfCell;
+  // rename variable cause it isn't self-linked after this point.
+
+  const next = prev.next;
+  cell.prev = prev;
+  cell.next = next;
+  prev.next = cell;
+  next.prev = cell;
+  // Not frozen!
+  return cell;
+};
+
+/**
+ * @param {DoublyLinkedCell} cell
+ * Must be a non-sigil part of a list, and therefore non-self-linked
+ */
+const spliceOut = cell => {
+  const { prev, next } = cell;
+  if (prev === cell || next === cell) {
+    throw TypeError('Expected non-self-linked cell');
+  }
+  prev.next = next;
+  next.prev = prev;
+  cell.prev = cell;
+  cell.next = cell;
+};
+
+/**
+ * The LRUCacheMap is used within the implementation of `assert` and so
+ * at a layer below SES or harden. Thus, we give it a `WeakMap`-like interface
+ * rather than a `WeakMapStore`-like interface. To work before `lockdown`,
+ * the implementation must use `freeze` manually, but still exhausively.
+ *
+ * It does not hold onto anything weakly. The only sense in which it is
+ * WeakMap-like is that it does not support enumeration.
+ *
+ * TODO: Make parameterized `Key` and `Value` template types
+ *
+ * @param {number} budget
+ * @returns {WeakMap<object,any>}
+ */
+export const makeLRUCacheMap = budget => {
+  if (!isSafeInteger(budget) || budget < 0) {
+    throw new TypeError('budget must be a safe non-negative integer number');
+  }
+  /** @type {Map<object, DoublyLinkedCell>} */
+  const map = new Map();
+  let count = 0; // count must remain <= budget
+  // As a sigil, `head` uniquely is not in the `map`.
+  const head = makeSelfCell(undefined, undefined);
+
+  const touchCell = key => {
+    const cell = map.get(key);
+    if (cell === undefined) {
+      return undefined;
+    }
+    // Becomes most recently used
+    spliceOut(cell);
+    spliceAfter(head, cell);
```

Comment:
constant time LRU. Is there a way to do that with an array?

Reply:
> constant time LRU. Is there a way to do that with an array?

Since the maximum size of the array would be bounded by `budget`, the answer to this question is _technically_ "yes". And it's possible that **`cache = [cache[i], ...cache.slice(0, i), ...cache.slice(i + 1)]`**[^1] or **`cache.splice(0, i + 1, cache[i], ...cache.slice(0, i))`**[^2] is faster than **`spliceOut(cell); spliceAfter(head, cell)`** in absolute terms for the budget we set even though it scales in a worse algorithmic time complexity class, or that it's slower but the difference doesn't matter so we could still go with simpler code.

[^1]: suitably protected against environment manipulation, e.g. `cache = [cache[i], ...arrayIterator(arraySlice(cache, 0, i)), ...arrayIterator(arraySlice(cache, i + 1))]`
[^2]: likewise (e.g., `arraySplice(cache, 0, i + 1, cache[i], ...arrayIterator(arraySlice(cache, 0, i)))`), and also with batching if `budget` can exceed maximum argument count

Reply:
I agree that is possible. Since it is a fully encapsulated decision, if measurements justify a change, we can always change it later.

Resolving.